content/docs/specs/remote_write_spec_2_0.md

```diff
@@ -0,0 +1,443 @@
+
+  However, Remote-Write 2.0 specification solves an important transactionality problem for [the classic histogram buckets](https://docs.google.com/document/d/1mpcSWH1B82q-BtJza-eJ8xMLlKt6EJ9oFGH325vtY1Q/edit#heading=h.ueg7q07wymku). This is done thanks to the native histograms supporting custom bucket-ing possible with the `io.prometheus.write.v2.Request` wire format. Sender might translate all classic histograms to native histograms this way, but it's out of this specification to mandate this. However, for this reason, Receiver MAY ignore certain metric types (e.g. classic histograms).
+
+* **Alternative wire formats**. The OpenTelemetry community has shown the validity of Apache Arrow (and potentially other columnar formats) for over-wire data transfer with their OTLP protocol. We would like to do experiments to confirm the compatibility of a similar format with Prometheus’ data model and include benchmarks of any resource usage changes. We would potentially maintain both a protobuf and columnar format long term for compatibility reasons and use our content negotiation to add different proto messages for this purpose.
+
+* **Global symbols**. Pre-defined string dictionary for interning The protocol could pre-define a static dictionary of ref->symbol that includes strings that are considered common, e.g. “namespace”, “le”, “job”, “seconds”, “bytes”, etc. Sender and refer to these without the need to include them in the request’s symbols table. This dictionary could incrementally grow with a minor version releases of this protocol.
+
+## Related
+
+### FAQ
+
+**Why did you not use gRPC?**
+Because 1.0 protocol does not use gRPC, breaking it would increase friction in the adoption. See 1.0 [reason](./remote_write_spec.md#faq).
+
+**Why not stream protobuf messages?**
+If you use persistent HTTP/1.1 connections, they are pretty close to streaming. Of course headers have to be re-sent, but yes that is less expensive than a new TCP set up.
+
+**Why do we send samples in order?**
+The in-order constraint comes from the encoding we use for time series data in Prometheus, the implementation of which is append-only. It is possible to remove this constraint, for instance by buffering samples and reordering them before encoding.
+
+**How can we parallelise requests with the in-order constraint?**
+Samples must be in-order _for a given series_. Remote-Write requests can be sent in parallel as long as they are for different series. In Prometheus, we shard the samples by their labels into separate queues, and then writes happen sequentially in each queue. This guarantees samples for the same series are delivered in order, but samples for different series are sent in parallel - and potentially "out of order" between different series.
```

Comment:
Is this up to date given that Prometheus supports out of order now? I know we still want in-order within a single request, I'm not saying we change the spec, but maybe these answers in the FAQ should be updated.

Is it for performance reasons maybe?

Reply:
Feel free to propose changes. I personally would avoid using OOO in Prometheus because of efficiency reasons.. and my lack of experience of using it (: ..but it's better and better and viable for "rare" cases.

Still I think the intention of this question is to "how do I send LONG back log of samples faster". Current wording (probably too much Prometheus details here which can get obsolete easily) suggests you NEVER send not-in-order samples (not allowed), but also NEVER send not-in-order samples in different TimeSeries for same series (spec does NOT prohibits this, but also don't recommends). You are right that technically the latter is an option for some systems e.g Prometheus with OOO enabled, if not common I guess. We could mention that "ensure Receiver supports OOO as it's not guaranteed/nor blocked by spec and then you can do it e.g. in Prometheus do XYZ"

Feel free to propose changes here, we could share something like above. 🤗

Reply:
How about... ?
```suggestion
**Why do we send samples in order?**
The in-order constraint comes from the encoding we use for time series data in Prometheus, the implementation of which is optimized for append-only workloads. Receiver may support out-of-order ingestion, though it is not guaranteed nor blocked by the specification.

**How can we parallelise requests with the in-order constraint?**
Samples must be in-order _for a given series_. Even if Receiver does not support out-of-order ingestion, Remote-Write requests can be sent in parallel as long as they are for different series. In Prometheus, we shard the samples by their labels into separate queues, and then writes happen sequentially in each queue. This guarantees samples for the same series are delivered in order, but samples for different series are sent in parallel - and potentially "out of order" between different series.
```

LMK what you think!

Reply:
Thanks! I added something as you proposed, with minor changes for clarity PTAL:

```
**Why do we send samples in order?**
The in-order constraint comes from the encoding we use for time series data in Prometheus, the implementation of which is optimized for append-only workloads. However, this requirement is also shared across many other databases and vendors in the ecosystem. In fact, [Prometheus with OOO feature enabled](https://youtu.be/qYsycK3nTSQ?t=1321), allows out-of-order writes, but with the performance penalty, thus reserved for rare events. To sum up, Receiver may support out-of-order ingestion, though it is not permitted by the specification. In the future e.g. 2.x spec versions, we could extend content type to negotiate the out-of-order writes, if needed.

**How can we parallelise requests with the in-order constraint?**
Samples must be in-order _for a given series_. However, even if Receiver does not support out-of-order ingestion, the Remote-Write requests can be sent in parallel as long as they are for different series. Prometheus shards the samples by their labels into separate queues, and then writes happen sequentially in each queue. This guarantees samples for the same series are delivered in order, but samples for different series are sent in parallel - and potentially "out of order" between different series.
```